File: packages/kotlin/src/main/kotlin/com/natsuneko/catalyst/models/EgeriaUser.kt
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

package com.natsuneko.catalyst.models

import kotlinx.serialization.Serializable

/**
 * User information
 */
@Serializable
data class EgeriaUser(
    val id: String,
    val screenName: String,
    val displayName: String,
    val profile: EgeriaUserProfile? = null,
    val profileEmoji: EgeriaUserProfileEmoji? = null
)

/**
 * The current user's profile emoji.
 *
 * This is a flattened representation of the `anyOf` union defined by the spec, discriminated
 * by [type]:
 * - `"standard"` populates [value] and [imageUrl]
 * - `"custom"` populates [id], [shortcode], [displayName], [imageUrl], [width] and [height]
 */
@Serializable
data class EgeriaUserProfileEmoji(
    val type: String,
    val value: String? = null,
    val id: String? = null,
    val shortcode: String? = null,
    val displayName: String? = null,
    val imageUrl: String? = null,
    val width: Double? = null,
    val height: Double? = null
)

@Serializable
data class EgeriaUserWrapper(val user: EgeriaUser)

/**
 * Wrapper for the `{"users": [...]}` response shape
 */
@Serializable
internal data class EgeriaUsersWrapper(val users: List<EgeriaUser>)

File: packages/swift/Sources/CatalystSwift/Models/EgeriaUser.swift
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

/// The profile emoji shown alongside a user's name.
///
/// This flattens the spec's `anyOf` union, discriminated by `type`:
/// - `"standard"` populates `value` and `imageUrl`.
/// - `"custom"` populates `id`, `shortcode`, `displayName`, `imageUrl`, `width` and `height`.
public struct EgeriaUserProfileEmoji: Decodable, Sendable, Equatable, Hashable {
  public let type: String
  public let value: String?
  public let id: String?
  public let shortcode: String?
  public let displayName: String?
  public let imageUrl: String?
  public let width: Double?
  public let height: Double?
}

public struct EgeriaUser: Decodable, Sendable, Equatable, Hashable {
  public let id: String
  public let screenName: String
  public let displayName: String
  public let profile: EgeriaUserProfile?
  public let profileEmoji: EgeriaUserProfileEmoji?
}

public struct EgeriaUserWrapper: Decodable, Sendable {
  public let user: EgeriaUser
}
